File: flutter-hms-analytics/ios/Classes/HMSAnalyticsMethodCallHandler.swift
/*
    Copyright 2020-2022. Huawei Technologies Co., Ltd. All rights reserved.

    Licensed under the Apache License, Version 2.0 (the "License")
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

import Flutter
import UIKit
import Foundation
import HiAnalytics

public class HMSAnalyticsMethodCallHandler: NSObject, FlutterPlugin {
    private lazy var analytics: Analytics = {
        Analytics.init()
    }()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.huawei.hms.flutter.analytics", binaryMessenger: registrar.messenger())
        let instance = HMSAnalyticsMethodCallHandler()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            return
        }
        let method = Methods.init()

        switch call.method {
        case method.GET_INSTANCE:
            guard let routePolicy = args["routePolicy"] as? String else { return }
            analytics.getInstance(routePolicy, resolve: result)
        case method.CLEAR_CACHED_DATA:
            analytics.clearCachedData(result)
        case method.SET_ANALYTICS_ENABLED:
            guard let enabled = args["enabled"] as? Bool else { return }
            analytics.setAnalyticsEnabled(enabled, resolve: result)
        case method.GET_AAID:
            analytics.aaid(result)
        case method.GET_USER_PROFILES:
            guard let predefined = args["predefined"] as? Bool else { return }
            analytics.userProfiles(predefined, resolve: result)
        case method.SET_SESSIONS_DURATION:
            guard let milisecond = args["duration"] as? Int else { return }
            analytics.setSessionDuration(milisecond, resolve: result)
        case method.SET_USER_PROFILE:
            guard let name = args["key"] as? String else { return }
            guard let value = args["value"] as? String else { return }
            analytics.setUserProfile(name, value: value, resolve: result)
        case method.ON_EVENT:
            guard let name = args["eventId"] as? String else { return }
            guard let value = args["params"] as? NSDictionary else { return }
            analytics.onEvent(name, params: value, resolve: result)
        case method.SET_USER_ID:
            guard let userId = args["userId"] as? String else { return }
            analytics.setUserId(userId, resolve: result)
        case method.SET_REPORT_POLICIES:
            guard let policyType = args["policyType"] as? NSDictionary else { return }
            analytics.setReportPolicies(policyType, resolve: result)
        case method.SET_RESTRICTION_ENABLED:
            guard let bool = args["enabled"] as? Bool else { return }
            analytics.setRestrictionEnabled(bool, resolve: result)
        case method.IS_RESTRICTION_ENABLED:
            analytics.isRestrictionEnabled(result)
        case method.DELETE_USER_PROFILE:
            guard let name = args["key"] as? String else { return }
            analytics.deleteUserProfile(name, resolve: result)
        case method.DELETE_USER_ID:
            analytics.deleteUserId(resolve: result)
        case method.SET_MIN_ACTIVITY_SESSIONS:
            guard let interval = args["interval"] as? Int64 else { return }
            analytics.setMinActivitySession(interval, resolve: result)
        case method.SET_COLLECT_ADS_ID_ENABLED:
            guard let enabled = args["enabled"] as? Bool else { return }
            analytics.setCollectAdsIdEnabled(enabled, resolve: result)
        case method.ADD_DEFAULT_EVENT_PARAMS:
            let params = args["params"] as? [String: Any]
            analytics.addDefaultEventParams(params, resolve: result)
        default:
            result(FlutterError(code: "platformError", message: "Not supported on iOS platform", details: ""))
        }
    }

    struct Methods {
        let GET_INSTANCE = "getInstance"
        let CLEAR_CACHED_DATA = "clearCachedData"
        let SET_ANALYTICS_ENABLED = "setAnalyticsEnabled"
        let GET_AAID = "getAAID"
        let GET_USER_PROFILES = "getUserProfiles"
        let SET_SESSIONS_DURATION = "setSessionDuration"
        let SET_USER_ID = "setUserId"
        let SET_USER_PROFILE = "setUserProfile"
        let ON_EVENT = "onEvent"
        let SET_REPORT_POLICIES = "setReportPolicies"
        let SET_RESTRICTION_ENABLED = "setRestrictionEnabled"
        let IS_RESTRICTION_ENABLED = "isRestrictionEnabled"
        let DELETE_USER_PROFILE = "deleteUserProfile"
        let DELETE_USER_ID = "deleteUserId"
        let SET_MIN_ACTIVITY_SESSIONS = "setMinActivitySessions"
        let SET_COLLECT_ADS_ID_ENABLED = "setCollectAdsIdEnabled"
        let ADD_DEFAULT_EVENT_PARAMS = "addDefaultEventParams"
    }
}

File: flutter-hms-analytics/android/src/main/java/com/huawei/hms/flutter/analytics/handler/HMSAnalyticsMethodCallHandler.java
/*
 * Copyright 2020-2024. Huawei Technologies Co., Ltd. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.hms.flutter.analytics.handler;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import com.huawei.hms.flutter.analytics.HMSAnalyticsModule;
import com.huawei.hms.flutter.analytics.logger.HMSLogger;

import java.lang.ref.WeakReference;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class HMSAnalyticsMethodCallHandler implements MethodCallHandler {
    private static final String TAG = HMSAnalyticsMethodCallHandler.class.getSimpleName();
    private HMSAnalyticsModule analyticsModule;

    // Weak Context Instance
    private final WeakReference<Context> weakContext;

    public HMSAnalyticsMethodCallHandler(HMSAnalyticsModule module, WeakReference<Context> weakContext) {
        this.analyticsModule = module;
        this.weakContext = weakContext;
    }

    public HMSAnalyticsMethodCallHandler(WeakReference<Context> weakContext) {
        this.weakContext = weakContext;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall methodCall, @NonNull Result result) {
        Log.i(TAG, " Running method : " + methodCall.method);
        HMSLogger.getInstance(weakContext.get()).startMethodExecutionTimer(methodCall.method);

        if (analyticsModule == null) {
            analyticsModule = new HMSAnalyticsModule(weakContext);
        }
        switch (Methods.valueOf(methodCall.method)) {
            case clearCachedData:
                analyticsModule.clearCachedData(result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setAnalyticsEnabled:
                analyticsModule.setAnalyticsEnabled(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case getAAID:
                analyticsModule.getAAID(result);
                break;
            case getUserProfiles:
                analyticsModule.getUserProfiles(methodCall, result);
                break;
            case pageStart:
                analyticsModule.pageStart(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case pageEnd:
                analyticsModule.pageEnd(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setMinActivitySessions:
                analyticsModule.setMinActivitySessions(methodCall, result);
                break;
            case setPushToken:
                analyticsModule.setPushToken(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setSessionDuration:
                analyticsModule.setSessionDuration(methodCall, result);
                break;
            case setUserId:
                analyticsModule.setUserId(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setUserProfile:
                analyticsModule.setUserProfile(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case onEvent:
                analyticsModule.onEvent(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case enableLog:
                analyticsModule.enableLog(result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case enableLogWithLevel:
                analyticsModule.enableLogWithLevel(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case enableLogger:
                analyticsModule.enableLogger(result);
                break;
            case disableLogger:
                analyticsModule.disableLogger(result);
                break;
            case getReportPolicyThreshold:
                analyticsModule.getReportPolicyThreshold(methodCall, result);
                break;
            case setReportPolicies:
                analyticsModule.setReportPolicies(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setRestrictionEnabled:
                analyticsModule.setRestrictionEnabled(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case isRestrictionEnabled:
                analyticsModule.isRestrictionEnabled(result);
                break;
            case setCollectAdsIdEnabled:
                analyticsModule.setCollectAdsIdEnabled(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case addDefaultEventParams:
                analyticsModule.addDefaultEventParams(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case deleteUserProfile:
                analyticsModule.deleteUserProfile(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case deleteUserId:
                analyticsModule.deleteUserId(result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setChannel:
                analyticsModule.setChannel(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setPropertyCollection:
                analyticsModule.setPropertyCollection(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case setCustomReferrer:
                analyticsModule.setCustomReferrer(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case getInstance:
                analyticsModule.getInstance(methodCall, result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            case getDataUploadSiteInfo:
                analyticsModule.getDataUploadSiteInfo(result);
                HMSLogger.getInstance(weakContext.get()).sendSingleEvent(methodCall.method);
                break;
            default:
                result.error("platformError", "Not supported on Android platform", "");
        }
    }

    private enum Methods {
        clearCachedData,
        setAnalyticsEnabled,
        getAAID,
        getUserProfiles,
        pageStart,
        pageEnd,
        setMinActivitySessions,
        setPushToken,
        setSessionDuration,
        setUserId,
        setUserProfile,
        onEvent,
        enableLog,
        enableLogWithLevel,
        enableLogger,
        disableLogger,
        getReportPolicyThreshold,
        setReportPolicies,
        setRestrictionEnabled,
        isRestrictionEnabled,
        setCollectAdsIdEnabled,
        addDefaultEventParams,
        deleteUserProfile,
        deleteUserId,
        setChannel,
        setPropertyCollection,
        setCustomReferrer,
        getInstance,
        getDataUploadSiteInfo
    }
}
